conn/physic/units_test.go

```diff
@@ -536,3 +537,1260 @@ func BenchmarkCelsiusFloatg(b *testing.B) {
+		postive  = false
+	)
+	tests := []struct {
+		s    string
+		want decimal
+		used int
+		err  bool
+	}{
+		{"123456789", decimal{"123456789", 0, postive}, 9, false},
+		{"1nM", decimal{"1", 0, postive}, 1, false},
+		{"2.2nM", decimal{"22", -1, postive}, 3, false},
+		{"12.5mA", decimal{"125", -1, postive}, 4, false},
+		{"-12.5mA", decimal{"125", -1, negative}, 5, false},
+		{"1.1.1", decimal{}, 0, true},
+		{"1ma1", decimal{"1", 0, postive}, 1, false},
+		{"-0.00001%rH", decimal{"1", -5, negative}, 8, false},
+		{"0.00001%rH", decimal{"1", -5, postive}, 7, false},
+		{"--1ma1", decimal{"1", 0, negative}, 3, false},
+		{"++100ma1", decimal{"1", 2, postive}, 5, false},
+		{"1.0", decimal{"1", 0, postive}, 3, false},
+		{"0.10001", decimal{"10001", -5, postive}, 7, false},
+		{"-0.10001", decimal{"10001", -5, negative}, 8, false},
+		{"%-0.10001", decimal{"10001", -5, negative}, 0, true},
+		{"1n", decimal{"1", 0, postive}, 1, false},
+		{"200n", decimal{"2", 2, postive}, 3, false},
+	}
+
+	for _, tt := range tests {
+		got, n, err := atod(tt.s)
+
+		if got != tt.want && !tt.err {
+			t.Errorf("got %v expected %v", got, tt.want)
+		}
+		if tt.err && err == nil {
+			t.Errorf("expected error %v but got nil", err)
+		}
+
+		if n != tt.used {
+			t.Errorf("expected to consume %d char but used %d", tt.used, n)
+		}
+	}
+}
+
+func TestDoti(t *testing.T) {
+	tests := []struct {
+		name string
+		d    decimal
+		want int64
+		err  bool
+	}{
+		{"123", decimal{"123", 0, false}, 123, false},
+		{"-123", decimal{"123", 0, true}, -123, false},
+		{"1230", decimal{"123", 1, false}, 1230, false},
+		{"-1230", decimal{"123", 1, true}, -1230, false},
+		{"1230", decimal{"123", 20, false}, 1230, true},
+		{"-1230", decimal{"123", 20, true}, -1230, true},
+		{"max", decimal{"9223372036854775807", 0, false}, 9223372036854775807, false},
+		{"-max", decimal{"9223372036854775807", 0, true}, -9223372036854775807, false},
+		{"max+1", decimal{"9223372036854775808", 0, true}, 0, true},
+		{"1a", decimal{"1a", 0, false}, 123, true},
+		{"2.7b", decimal{"2.7b", 0, true}, -123, true},
+		{"12", decimal{"123", -1, false}, 12, false},
+		{"-12", decimal{"123", -1, true}, -12, false},
+		{"123n", decimal{"123", 0, false}, 123, false},
+		{"max*10^1", decimal{"9223372036854775807", 1, false}, 9223372036854775807, true},
+		{"overflow", decimal{"9223372036854775807", 10, false}, 9223372036854775807, true},
+	}
+
+	for _, tt := range tests {
+		got, err := tt.d.dtoi(0)
+
+		if got != tt.want && !tt.err {
+			t.Errorf("got %v expected %v", got, tt.want)
+		}
+		if tt.err && err == nil {
+			t.Errorf("expected %v but got nil, %v", err, got)
+		}
+	}
+
+}
+
+func BenchmarkSetDistance(b *testing.B) {
+	var t Temperature
+	for i := 0; i < b.N; i++ {
+		t.Set("-337.2C")
+	}
+}
+
+func BenchmarkSetPower(b *testing.B) {
+	var t Power
+	for i := 0; i < b.N; i++ {
+		t.Set("-337.2w")
+	}
+}
+
+func BenchmarkDecimal(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		atod("-337.2m")
```

Comment:
You need to look at the result otherwise there's a risk of the compiler to elide part of the code. Often what I do is:
```
t := 0
for i := 0; i < b.N; i++ {
  _, t, _ = atod("-337.2m")
}
b.StopTimer()
fmt.Sprintf("%s", t)
```
just so the compiler cannot elide the calculation.